File: model-service/src/infrastructure/adapters/inbound/fastapi/schemas/claims.py
"""Pydantic schemas for claim extraction and synthesis endpoints.

This module defines request and response schemas for the /api/claims/* endpoints.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import Field

from src.infrastructure.adapters.inbound.fastapi.schemas.common import (
    ConfidenceScore,
    NonEmptyStr,
    PositiveInt,
    ProcessingTime,
    StrictBaseModel,
)


class ExtractedClaim(StrictBaseModel):
    """Single extracted claim with metadata.

    Attributes
    ----------
    text : str
        Claim text.
    sentence_index : int | None
        Index of source sentence (if sentence-based).
    char_start : int | None
        Character offset in summary text.
    char_end : int | None
        Character offset end in summary text.
    subclaims : list[ExtractedClaim]
        Nested subclaims.
    confidence : float
        Model confidence in claim extraction.
    claim_type : str | None
        Semantic type of claim.
    """

    text: NonEmptyStr = Field(..., description="Claim text")
    sentence_index: int | None = Field(
        default=None, ge=0, description="Index of source sentence (if sentence-based)"
    )
    char_start: int | None = Field(
        default=None, ge=0, description="Character offset in summary text"
    )
    char_end: int | None = Field(
        default=None, ge=0, description="Character offset end in summary text"
    )
    subclaims: list[ExtractedClaim] = Field(
        default_factory=list, description="Nested subclaims"
    )
    confidence: ConfidenceScore = Field(
        ..., description="Model confidence in claim extraction"
    )
    claim_type: str | None = Field(default=None, description="Semantic type of claim")


class ClaimExtractionRequest(StrictBaseModel):
    """Request model for claim extraction endpoint.

    Attributes
    ----------
    summary_id : str
        Unique identifier for the summary.
    summary_text : str
        Full summary text to extract claims from.
    sentences : list[str] | None
        Pre-split sentences (optional).
    annotations : list[dict[str, Any]] | None
        Annotation data for context.
    ontology_types : list[dict[str, Any]] | None
        Ontology type definitions for context.
    ontology_glosses : dict[str, str] | None
        Map of type ID to gloss text.
    extraction_strategy : str
        Strategy for extracting claims.
    max_claims : int
        Maximum number of claims to extract.
    min_confidence : float
        Minimum confidence threshold for claims.
    """

    summary_id: NonEmptyStr = Field(..., description="Unique identifier for the summary")
    summary_text: NonEmptyStr = Field(..., description="Full summary text to extract claims from")
    sentences: list[str] | None = Field(
        default=None, description="Pre-split sentences (optional, will split if not provided)"
    )

    # Optional context sources
    annotations: list[dict[str, Any]] | None = Field(
        default=None, description="Annotation data for context (object names, times, etc.)"
    )
    ontology_types: list[dict[str, Any]] | None = Field(
        default=None, description="Ontology type definitions for context"
    )
    ontology_glosses: dict[str, str] | None = Field(
        default=None, description="Map of type ID to gloss text"
    )

    # Extraction configuration
    extraction_strategy: Literal["sentence-based", "semantic-units", "hierarchical"] = Field(
        default="sentence-based", description="Strategy for extracting claims"
    )
    max_claims: int = Field(
        default=50, ge=1, le=200, description="Maximum number of claims to extract"
    )
    min_confidence: ConfidenceScore = Field(
        default=0.5, description="Minimum confidence threshold for claims"
    )


class ClaimExtractionResponse(StrictBaseModel):
    """Response model for claim extraction endpoint.

    Attributes
    ----------
    summary_id : str
        Summary identifier.
    claims : list[ExtractedClaim]
        Extracted claims.
    model_used : str
        LLM model used for extraction.
    processing_time : float
        Processing time in seconds.
    """

    summary_id: NonEmptyStr = Field(..., description="Summary identifier")
    claims: list[ExtractedClaim] = Field(..., description="Extracted claims")
    model_used: NonEmptyStr = Field(..., description="LLM model used for extraction")
    processing_time: ProcessingTime = Field(..., description="Processing time in seconds")


class ClaimSource(StrictBaseModel):
    """Source of claims for synthesis (single video or collection).

    Attributes
    ----------
    source_id : str
        Video ID or collection ID.
    source_type : str
        Type of source (video or collection).
    claims : list[dict[str, Any]]
        Hierarchical claim structure.
    metadata : dict[str, Any] | None
        Source metadata (video title, date, etc.).
    """

    source_id: NonEmptyStr = Field(..., description="Video ID or collection ID")
    source_type: Literal["video", "collection"] = Field(..., description="Type of source")
    claims: list[dict[str, Any]] = Field(..., description="Hierarchical claim structure")
    metadata: dict[str, Any] | None = Field(
        default=None, description="Source metadata (video title, date, etc.)"
    )


class ClaimRelationship(StrictBaseModel):
    """Relationship between claims across sources.

    Attributes
    ----------
    source_claim_id : str
        Source claim ID.
    target_claim_id : str
        Target claim ID.
    relation_type : str
        Type of relationship.
    confidence : float
        Confidence score.
    notes : str | None
        Optional notes.
    """

    source_claim_id: NonEmptyStr = Field(..., description="Source claim ID")
    target_claim_id: NonEmptyStr = Field(..., description="Target claim ID")
    relation_type: Literal[
        "supports",
        "conflicts_with",
        "contradicts",
        "refines",
        "generalizes",
        "duplicates",
    ] = Field(..., description="Type of relationship")
    confidence: ConfidenceScore = Field(default=0.8, description="Confidence score")
    notes: str | None = Field(default=None, description="Optional notes")


class SummarySynthesisRequest(StrictBaseModel):
    """Request model for summary synthesis endpoint.

    Attributes
    ----------
    summary_id : str
        Target summary identifier.
    claim_sources : list[ClaimSource]
        Claim hierarchies from one or more sources.
    claim_relations : list[ClaimRelationship] | None
        Relationships between claims.
    ontology_context : dict[str, Any] | None
        Ontology types and glosses for references.
    persona_context : dict[str, Any] | None
        Persona information for perspective.
    synthesis_strategy : str
        Strategy for organizing summary.
    max_length : int
        Maximum summary length in words.
    include_conflicts : bool
        Explicitly mention informational conflicts.
    include_citations : bool
        Include inline citations to source claims.
    """

    summary_id: NonEmptyStr = Field(..., description="Target summary identifier")

    # Input sources (single or multiple)
    claim_sources: list[ClaimSource] = Field(
        ..., min_length=1, description="Claim hierarchies from one or more sources"
    )

    # Inter-claim relationships
    claim_relations: list[ClaimRelationship] | None = Field(
        default=None, description="Relationships between claims (conflicts, support, etc.)"
    )

    # Context for synthesis
    ontology_context: dict[str, Any] | None = Field(
        default=None, description="Ontology types and glosses for # references"
    )
    persona_context: dict[str, Any] | None = Field(
        default=None, description="Persona information for perspective"
    )

    # Synthesis configuration
    synthesis_strategy: Literal[
        "hierarchical",
        "chronological",
        "narrative",
        "analytical",
    ] = Field(default="hierarchical", description="Strategy for organizing summary")

    max_length: int = Field(
        default=500, ge=100, le=2000, description="Maximum summary length in words"
    )

    include_conflicts: bool = Field(
        default=True, description="Explicitly mention informational conflicts"
    )

    include_citations: bool = Field(
        default=False, description="Include inline citations to source claims"
    )


class SummarySynthesisResponse(StrictBaseModel):
    """Response model for summary synthesis endpoint.

    Attributes
    ----------
    summary_id : str
        Summary identifier.
    summary_gloss : list[dict[str, Any]]
        Generated summary as GlossItem array with references.
    model_used : str
        LLM model used for synthesis.
    processing_time : float
        Processing time in seconds.
    claims_used : int
        Total claims synthesized.
    synthesis_metadata : dict[str, Any]
        Metadata about synthesis.
    """

    summary_id: NonEmptyStr = Field(..., description="Summary identifier")
    summary_gloss: list[dict[str, Any]] = Field(
        ..., description="Generated summary as GlossItem array with # and @ references"
    )
    model_used: NonEmptyStr = Field(..., description="LLM model used for synthesis")
    processing_time: ProcessingTime = Field(..., description="Processing time in seconds")
    claims_used: int = Field(..., ge=0, description="Total claims synthesized")
    synthesis_metadata: dict[str, Any] = Field(
        ..., description="Metadata about synthesis (strategy, conflicts, etc.)"
    )

File: annotation-tool/src/models/claims.ts
import type { GlossItem } from './gloss'

/**
 * @description Types of claim sources.
 * Identifies who or what is making a claim.
 */
export type ClaimerType =
  /** Single entity from world state making the claim */
  | 'entity'
  /** Entity type from ontology (generic claimer) */
  | 'entity_type'
  /** Video creator explicitly asserts this claim */
  | 'author'
  /** Mix of text, types, and objects */
  | 'mixed'

/**
 * @description Strategies for extracting claims from summaries.
 * Determines how claims are identified and structured.
 */
export type ExtractionStrategy =
  /** One claim per sentence, decompose into subclaims */
  | 'sentence-based'
  /** Extract claims from semantic chunks */
  | 'semantic-units'
  /** Top-down hierarchical decomposition */
  | 'hierarchical'
  /** User-created claim (not auto-extracted) */
  | 'manual'

/**
 * @interface ClaimSpan
 * @description Span within claim text for discontiguous selections.
 * Used to highlight specific portions of claim text.
 */
export interface ClaimSpan {
  /** Starting character position (0-indexed) */
  charStart: number
  /** Ending character position (exclusive) */
  charEnd: number
}

/**
 * @interface ClaimTextSpan
 * @description Text span with optional sentence index.
 * Extends ClaimSpan with sentence-level information.
 */
export interface ClaimTextSpan extends ClaimSpan {
  /** Index of the sentence this span belongs to */
  sentenceIndex?: number
}

/**
 * @interface Claim
 * @description An atomic factual statement extracted from a summary.
 * Can be hierarchically decomposed into subclaims.
 *
 * @remarks
 * Claims are the fundamental unit of information extraction. They represent
 * statements that can be verified, attributed to a source, and related to
 * other claims. Claims support rich text with references to ontology types
 * and world objects using # @ ^ $ notation.
 *
 * @example
 * ```typescript
 * const claim: Claim = {
 *   id: 'claim-123',
 *   summaryId: 'summary-456',
 *   summaryType: 'video',
 *   text: 'The speaker discusses climate change.',
 *   gloss: [
 *     { type: 'text', content: 'The speaker discusses ' },
 *     { type: 'typeRef', content: 'climate change', refType: 'event' }
 *   ],
 *   claimerType: 'author',
 *   confidence: 0.9,
 *   extractionStrategy: 'sentence-based',
 *   createdAt: '2024-01-01T00:00:00Z',
 *   updatedAt: '2024-01-01T00:00:00Z'
 * };
 * ```
 */
export interface Claim {
  /** Unique identifier for the claim */
  id: string
  /** ID of the summary this claim was extracted from */
  summaryId: string
  /** Type of summary (video or collection) */
  summaryType: 'video' | 'collection'

  /** Plain text content of the claim */
  text: string
  /** Rich text content with optional references */
  gloss: GlossItem[]

  /** ID of parent claim for hierarchical decomposition */
  parentClaimId?: string
  /** Child claims (subclaims) */
  subclaims?: Claim[]

  /** Text spans mapping claim to source text (discontiguous) */
  textSpans?: ClaimTextSpan[]

  /** Type of entity making this claim */
  claimerType?: ClaimerType | null
  /** Rich text identifying who is making the claim */
  claimerGloss?: GlossItem[]
  /** How the claimer relates to the claim content */
  claimRelation?: GlossItem[]

  /** Event during which this claim was made */
  claimEventId?: string
  /** Time when this claim was made */
  claimTimeId?: string
  /** Location where this claim was made */
  claimLocationId?: string

  /** Confidence score for this claim (0-1) */
  confidence?: number
  /** Model used to extract this claim */
  modelUsed?: string
  /** Strategy used for extraction */
  extractionStrategy?: ExtractionStrategy

  /** Modality metadata - indicates what sources support the claim */
  /** Array of "speech" and/or "non-speech" - claim is based at least in part on audio */
  audio?: ('speech' | 'non-speech')[] | null
  /** Array of "text" and/or "non-text" - claim is based at least in part on non-audio video information */
  video?: ('text' | 'non-text')[] | null
  /** Array of "text" (caption metadata) and/or "non-text" (other metadata like location from .info.json) */
  metadata?: ('text' | 'non-text')[] | null
  /** Optional comment about this claim */
  comment?: string | null

  /** Relations where this claim is the source */
  sourceClaimRelations?: ClaimRelation[]
  /** Relations where this claim is the target */
  targetClaimRelations?: ClaimRelation[]

  /** ID of user who created this claim */
  createdBy?: string
  /** ISO 8601 timestamp of creation */
  createdAt: string
  /** ISO 8601 timestamp of last update */
  updatedAt: string
}

/**
 * @interface ClaimRelation
 * @description Relation between claims or claim spans.
 * Represents how claims relate to each other (e.g., supports, contradicts).
 */
export interface ClaimRelation {
  /** Unique identifier for this relation */
  id: string
  /** ID of the source claim */
  sourceClaimId: string
  /** ID of the target claim */
  targetClaimId: string
  /** ID of the relation type from ontology */
  relationTypeId: string

  /** Specific spans in source claim involved in relation */
  sourceSpans?: ClaimSpan[]
  /** Specific spans in target claim involved in relation */
  targetSpans?: ClaimSpan[]

  /** Confidence score for this relation (0-1) */
  confidence?: number
  /** User notes about this relation */
  notes?: string
  /** ID of user who created this relation */
  createdBy?: string
  /** ISO 8601 timestamp of creation */
  createdAt: string
  /** ISO 8601 timestamp of last update */
  updatedAt: string
}

/**
 * @interface ClaimExtractionConfig
 * @description Configuration for automated claim extraction.
 * Controls what sources to use, extraction strategy, and post-processing.
 */
export interface ClaimExtractionConfig {
  /**
   * Input sources for extraction.
   */
  inputSources: {
    /** Include the summary text as input */
    includeSummaryText: boolean
    /** Include annotation data as context */
    includeAnnotations: boolean
    /** Include ontology definitions as context */
    includeOntology: boolean
    /** How much ontology detail to include */
    ontologyDepth: 'names-only' | 'names-and-glosses' | 'full-definitions'
  }

  /** Strategy for identifying and structuring claims */
  extractionStrategy: ExtractionStrategy
  /** Maximum claims to extract per summary (default: 50) */
  maxClaimsPerSummary?: number
  /** Maximum subclaim nesting depth (default: 3) */
  maxSubclaimDepth?: number
  /** Minimum confidence threshold (default: 0.5) */
  minConfidence?: number

  /** Specific LLM model ID to use */
  modelId?: string

  /** Remove duplicate claims (default: true) */
  deduplicateClaims?: boolean
  /** Merge highly similar claims (default: false) */
  mergeSimilarClaims?: boolean
}

/**
 * @interface ClaimStructure
 * @description Denormalized claim structure stored in JSON field.
 * Contains all claims with metadata about the extraction.
 */
export interface ClaimStructure {
  /** Schema version (e.g., "1.0") */
  version: string
  /** Root claims with nested subclaims */
  claims: Claim[]
  /** Metadata about the extraction */
  metadata: {
    /** ISO 8601 timestamp of extraction */
    extractedAt: string
    /** Model used for extraction */
    modelUsed: string
    /** Configuration used for extraction */
    config: ClaimExtractionConfig
    /** Total number of root claims */
    totalClaims: number
    /** Total number of subclaims (all levels) */
    totalSubclaims: number
    /** Maximum nesting depth reached */
    maxDepth: number
  }
}

/**
 * @interface VideoSummaryWithClaims
 * @description VideoSummary extended with extracted claims.
 * Combines summary content with claim extraction results.
 */
export interface VideoSummaryWithClaims {
  /** Unique identifier for the summary */
  id: string
  /** ID of the video being summarized */
  videoId: string
  /** ID of the persona who wrote this summary */
  personaId: string
  /** Rich text summary with optional references */
  summary: GlossItem[]
  /** ISO 8601 timestamp of creation */
  createdAt: string
  /** ISO 8601 timestamp of last update */
  updatedAt: string
  /** ID of the user who created this summary */
  createdBy?: string
  /** Extracted claims (denormalized) */
  claims?: Claim[]
  /** Full claim structure with metadata */
  claimsJson?: ClaimStructure
  /** Version of claim extraction schema */
  claimsVersion?: string
  /** When claims were extracted */
  claimsExtractedAt?: string
}

/**
 * @interface ExtractClaimsRequest
 * @description Request to extract claims from a summary.
 */
export interface ExtractClaimsRequest {
  /** ID of the summary to extract from */
  summaryId: string
  /** Type of summary (video or collection) */
  summaryType: 'video' | 'collection'
  /** Extraction configuration */
  config: ClaimExtractionConfig
}

/**
 * @interface ExtractClaimsResponse
 * @description Response from claim extraction request.
 * Returns job information for async processing.
 */
export interface ExtractClaimsResponse {
  /** Unique job identifier */
  jobId: string
  /** Current job status */
  status: 'queued' | 'processing' | 'completed' | 'failed'
  /** ID of the summary being processed */
  summaryId: string
  /** Type of summary being processed */
  summaryType: string
}

/**
 * @interface ClaimExtractionJobStatus
 * @description Status of a claim extraction job.
 * Used for polling job progress.
 */
export interface ClaimExtractionJobStatus {
  /** Unique job identifier */
  jobId: string
  /** Current job status */
  status: 'queued' | 'processing' | 'completed' | 'failed'
  /** Progress percentage (0-100) */
  progress?: number
  /** Extraction result (when completed) */
  result?: ClaimStructure
  /** Error message (when failed) */
  error?: string
}

/**
 * @interface CreateClaimRequest
 * @description Request to create a claim manually.
 */
export interface CreateClaimRequest {
  /** ID of the summary this claim belongs to */
  summaryId: string
  /** Type of summary */
  summaryType: 'video' | 'collection'
  /** Plain text content of the claim */
  text: string
  /** Rich text content with references */
  gloss?: GlossItem[]
  /** Parent claim ID for hierarchical structure */
  parentClaimId?: string
  /** Text spans in source summary */
  textSpans?: ClaimTextSpan[]
  /** Type of claimer */
  claimerType?: ClaimerType | null
  /** Rich text claimer identification */
  claimerGloss?: GlossItem[]
  /** How claimer relates to claim */
  claimRelation?: GlossItem[]
  /** Associated event ID */
  claimEventId?: string
  /** Associated time ID */
  claimTimeId?: string
  /** Associated location ID */
  claimLocationId?: string
  /** Confidence score */
  confidence?: number
  /** Modality metadata */
  audio?: ('speech' | 'non-speech')[] | null
  video?: ('text' | 'non-text')[] | null
  metadata?: ('text' | 'non-text')[] | null
  /** Optional comment about this claim */
  comment?: string | null
}

/**
 * @interface UpdateClaimRequest
 * @description Request to update an existing claim.
 * All fields are optional - only provided fields are updated.
 */
export interface UpdateClaimRequest {
  /** Updated plain text content */
  text?: string
  /** Updated rich text content */
  gloss?: GlossItem[]
  /** Updated text spans */
  textSpans?: ClaimTextSpan[]
  /** Updated claimer type */
  claimerType?: ClaimerType | null
  /** Updated claimer gloss */
  claimerGloss?: GlossItem[]
  /** Updated claim relation */
  claimRelation?: GlossItem[]
  /** Updated event ID */
  claimEventId?: string
  /** Updated time ID */
  claimTimeId?: string
  /** Updated location ID */
  claimLocationId?: string
  /** Updated confidence score */
  confidence?: number
  /** Updated modality metadata */
  audio?: ('speech' | 'non-speech')[] | null
  video?: ('text' | 'non-text')[] | null
  metadata?: ('text' | 'non-text')[] | null
  /** Updated comment */
  comment?: string | null
}
